File: bind/typescript/src/builtin/files.ts
import { mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import type { Hearing, WorldContext, WorldPart } from "../extension.js";
import { type Fact, isQuestion } from "../types.js";

/** The most bytes the World reads of one file, since a text a model cannot hold is no answer. */
const CAP = 524288;

/** The part of the files extension: a read and a write of the disk, each resolved where its chain stands, answered
 * with the plain data of the path and the content, or with a refusal. A path of a scheme is another ear's to
 * answer, and the door of an act that no longer lives is refused. */
export default function files(context: WorldContext): WorldPart {
  /** Whether a path is a door: its first part names an act of the life, so `./` before it reaches the file. */
  function* door(path: string): Hearing<boolean> {
    try {
      yield { verb: "get", args: [path.split("/")[0]] };
      return true;
    } catch {
      return false;
    }
  }
  function read(path: string): { path: string; content: string } {
    let info: ReturnType<typeof statSync>;
    try {
      info = statSync(path);
    } catch (error) {
      // A path that names nothing is said in plain words, which the model and the operator both read.
      if ((error as NodeJS.ErrnoException).code === "ENOENT") throw new Error(`There is no file at ${path}.`);
      throw error;
    }
    if (!info.isFile() || info.size > CAP)
      throw new Error(`Read needs a text file at most ${CAP} bytes: ${path}`);
    return {
      path,
      content: new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(readFileSync(path)),
    };
  }
  function write(path: string, content: string): { path: string; content: string } {
    let before = "";
    try {
      before = readFileSync(path, "utf8");
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
    }
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, content);
    context.change({ path, before, after: content });
    return { path, content: readFileSync(path, "utf8") };
  }
  return {
    *hears(fact: Fact) {
      const [kind, qid, , on, path, content] = fact;
      if ((kind !== "read" && kind !== "write") || !isQuestion(kind, qid) || typeof path !== "string") return;
      if (path.includes("://")) return;
      let value: unknown;
      try {
        if (yield* door(path))
          throw new Error(
            `${path} is the door of nothing that ${kind === "read" ? "lives" : "takes a word"}`,
          );
        const at = context.at(yield* context.where(String(on)), path);
        value = kind === "read" ? read(at) : write(at, String(content));
      } catch (error) {
        value = context.refused(error instanceof Error ? error.message : String(error));
      }
      yield ["done", qid, value];
    },
  };
}

File: src/world/files.rs
//! The files of the machine: a read and a write of a path that nobody of the engine serves.

use std::{fs, io::ErrorKind, path::Path};

use super::{here, resolved};
use crate::{
  ear::{Ear, ear, hear, say},
  fact::Fact,
  value::{Fault, Text},
};

/// The largest text a read gives, in bytes.
const LARGEST: u64 = 524_288;

/// The ear of the files: it answers a read with the text at the path and a write with the text as it stands after,
/// each resolved against the working directory of the chain, or with the refusal.
pub fn files() -> Box<dyn Ear> {
  ear(|co, _| async move {
    loop {
      let a = hear(&co).await;
      if !a.question() || !matches!(a.kind(), "read" | "write") {
        continue;
      }
      let answer = match served(&co, &a).await {
        Ok(text) => text.object(),
        Err(fault) => fault.object(),
      };
      say(&co, Fact::says("done", a.about(), [answer])).await;
    }
  })
}

/// What a read or a write came to.
async fn served(co: &crate::ear::Co, a: &Fact) -> Result<Text, Fault> {
  let (path, content) = match a.kind() {
    "read" => (a.word(1).and_then(|one| one.as_str().map(str::to_owned)).unwrap_or_default(), None),
    _ => {
      let text =
        a.word(1).and_then(Text::of).ok_or_else(|| Fault::refused("a write gives a text"))?;
      (text.path, Some(text.content))
    }
  };
  if path.contains("://") {
    return Err(Fault::refused(format!("No file at {path}.")));
  }
  let at = resolved(&here(co, a.on()).await?, &path);
  if let Some(content) = content {
    if let Some(parent) = at.parent() {
      fs::create_dir_all(parent).map_err(|no| failed(&no, &at))?;
    }
    fs::write(&at, content).map_err(|no| failed(&no, &at))?;
  }
  read(&at)
}

/// The text at a path: a file of utf-8 no larger than a read gives.
fn read(at: &Path) -> Result<Text, Fault> {
  let shown = at.display();
  let info = fs::metadata(at).map_err(|no| failed(&no, at))?;
  if !info.is_file() || info.len() > LARGEST {
    return Err(Fault::refused(format!("Read needs a text file at most {LARGEST} bytes: {shown}")));
  }
  let bytes = fs::read(at).map_err(|no| failed(&no, at))?;
  let content = String::from_utf8(bytes)
    .map_err(|_| Fault::refused(format!("Read needs a text file in utf-8: {shown}")))?;
  Ok(Text::new(shown.to_string(), content))
}

/// What the machine refused, in plain words, which the model and the operator both read.
fn failed(no: &std::io::Error, at: &Path) -> Fault {
  match no.kind() {
    ErrorKind::NotFound => Fault::refused(format!("There is no file at {}.", at.display())),
    _ => Fault::refused(format!("{no}: {}", at.display())),
  }
}
